rs/ic_os/build_tools/icsum/src/main.rs

```diff
@@ -0,0 +1,319 @@
+        Ok(result) => Ok(Some(result as u64)),
+        Err(Errno::ENXIO) => Ok(None),
+        Err(err) => Err(err.into()),
+    }
+}
+
+/// See [seek_hole] for documentation.
+fn seek_data(file: &mut File, from: u64) -> std::io::Result<Option<u64>> {
+    // NOTE: u64 does not fully fit in i64, but i64::MAX is 9_223_372_036_854_775_807
+    // (8 exbibytes) and we won't be seeing files this large any time soon.
+    let unsized_from = from.try_into().unwrap();
+
+    match nix::unistd::lseek(file.as_raw_fd(), unsized_from, Whence::SeekData) {
+        Ok(result) => Ok(Some(result as u64)),
+        Err(Errno::ENXIO) => Ok(None),
+        Err(err) => Err(err.into()),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::rngs::StdRng;
+    use rand::{RngCore, SeedableRng};
+    use std::io::Write;
+    use tempfile::tempfile;
+
+    mod iterate_blocks {
+        use super::*;
+
+        #[test]
+        /// This test ensures that the test environment supports sparse files.
+        /// Otherwise, the rest of the test cases would be useless.
+        fn file_system_supports_sparse_files() {
+            let mut file = tempfile().unwrap();
+            file.seek(SeekFrom::Start(BLOCK_LEN_U64)).unwrap();
+            writeln!(&mut file, "hello").unwrap();
+
+            file.rewind().unwrap();
+
+            let mut holes = 0;
+            iterate_blocks(&mut file, |block| {
+                if block == Block::Hole {
+                    holes += 1;
+                }
+            })
+            .unwrap();
+
+            assert!(
+                holes >= 1,
+                "Expected at least 1 hole, filesystem likely doesn't support sparse files."
+            );
+        }
+
+        #[test]
+        fn no_holes_prime_bytes() {
+            let mut file = tempfile().unwrap();
+            let mut rng = StdRng::seed_from_u64(0);
+            let mut data = vec![0u8; 26205239];
+            rng.fill_bytes(&mut data);
+
+            file.write_all(&data).unwrap();
+
+            verify_file(file);
+        }
+
+        #[test]
+        fn no_holes_32mb() {
+            let mut file = tempfile().unwrap();
+            let mut rng = StdRng::seed_from_u64(0);
+            let mut data = vec![0u8; 32 * 1024 * 1024];
+            rng.fill_bytes(&mut data);
+
+            file.write_all(&data).unwrap();
+
+            verify_file(file);
+        }
+
+        #[test]
+        fn prime_holes_and_sizes() {
+            let mut file = tempfile().unwrap();
+            let mut rng = StdRng::seed_from_u64(0);
+            let mut data = vec![0u8; 26205239];
+            rng.fill_bytes(&mut data);
+
+            file.seek(SeekFrom::Current(9590653)).unwrap();
+            file.write_all(&data).unwrap();
+
+            file.seek(SeekFrom::Current(26205239)).unwrap();
+            file.write_all(&data).unwrap();
+
+            verify_file(file);
+        }
+
+        #[test]
+        fn _8mb_holes_and_sizes() {
```

Comment:
Underscore prefixes are [sometimes special](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=%23%5Bmust_use%5D%0Astruct+MustUse%3B%0A%0Afn+main%28%29+%7B%0A++++let+_okay+%3D+make_must_use%28%29%3B%0A++++%0A++++let+not_okay+%3D+make_must_use%28%29%3B%0A%7D%0A%0Afn+make_must_use%28%29+-%3E+MustUse+%7B%0A++++MustUse%0A%7D) in Rust.

I don't think they have any effect on fns, but could we avoid the name anyway?